src/tests/integration/D3MCompound.t.sol

```diff
@@ -0,0 +1,1234 @@
+                // Keep going after restoring the original value
+                hevm.store(
+                    address(token),
+                    keccak256(abi.encode(address(this), uint256(i))),
+                    prevValue
+                );
+            }
+        }
+
+        // We have failed if we reach here
+        assertTrue(false);
+    }
+
+    function assertEqRounding(uint256 _a, uint256 _b) internal {
+        _assertEqApprox(_a, _b, 10 ** 10);
+    }
+
+    function assertEqCdai(uint256 _a, uint256 _b) internal {
+        _assertEqApprox(_a, _b, 1);
+    }
+
+    function assertEqVatDai(uint256 _a, uint256 _b) internal {
+        _assertEqApprox(_a, _b, (10 ** 10) * RAY);
+    }
+
+    function assertEqInterest(uint256 _a, uint256 _b) internal {
+        _assertEqApprox(_a, _b, INTEREST_RATE_TOLERANCE);
+    }
+
+    function _assertEqApprox(uint256 _a, uint256 _b, uint256 _tolerance) internal {
+        uint256 a = _a;
+        uint256 b = _b;
+        if (a < b) {
+            uint256 tmp = a;
+            a = b;
+            b = tmp;
+        }
+        if (a - b > _tolerance) {
+            emit log_bytes32("Error: Wrong `uint' value");
+            emit log_named_uint("  Expected", _b);
+            emit log_named_uint("    Actual", _a);
+            fail();
+        }
+    }
+
+    function assertEqRoundingAgainst(uint256 _a, uint256 _b) internal {
+        uint256 a = _a;
+        uint256 b = _b;
+        if (a < b) {
+            uint256 tmp = a;
+            a = b;
+            b = tmp;
+        }
+        if (a - b > 1) {
+            emit log_bytes32("Error: Wrong `uint' value");
+            emit log_named_decimal_uint("  Expected", _b, 27);
+            emit log_named_decimal_uint("    Actual", _a, 27);
+            fail();
+        }
+    }
+
+    function getBorrowRate() public view returns (uint256 borrowRate) {
+        borrowRate = cDai.borrowRatePerBlock();
+    }
+
+    // Set the borrow rate to a relative percent to what it currently is
+    function _setRelBorrowTarget(uint256 deltaBPS) internal returns (uint256 targetBorrowRate) {
+        targetBorrowRate = getBorrowRate() * deltaBPS / 10000;
+        d3mCompoundPlan.file("barb", targetBorrowRate);
+        d3mHub.exec(ilk);
+    }
+
+    function test_target_decrease() public {
+        uint256 targetBorrowRate = _setRelBorrowTarget(7500);
+        d3mHub.reap(ilk);     // Clear out interest to get rid of rounding errors
+        assertEqInterest(getBorrowRate(), targetBorrowRate);
+
+        uint256 amountSupplied = cDai.balanceOfUnderlying(address(d3mCompoundPool));
+
+        assertTrue(amountSupplied > 0);
+        (uint256 ink, uint256 art) = vat.urns(ilk, address(d3mCompoundPool));
+        assertEqRounding(ink, amountSupplied);
+        assertEqRounding(art, amountSupplied);
+
+        assertEq(vat.gem(ilk, address(d3mCompoundPool)), 0);
+        assertEq(vat.dai(address(d3mHub)), 0);
+    }
+
+    function test_target_increase() public {
```

Comment:
Maybe we should cache the `amountSupplied` before raising the rate to prove after doing it will be lower.

Reply:
Handled in `eba23f7018`